### ISSUE TITLE
[Bug] cpu memory leak while using GPU with variable length ops.

### ISSUE THREAD

**Issue by unrealwill** [tensorflow/tensorflow#21348]
```
import tensorflow as tf
import numpy as np
import gc

def NotBuggy():
    with tf.device("/gpu:0"):
        x = tf.placeholder(dtype=tf.float32, shape=(1000,1))
        col = tf.placeholder(dtype=tf.int32, shape=(None,1))

        p0 = tf.gather_nd( x,col)
        p1 = tf.gather_nd( x,col+1)
        diff = p0 - p1
    sess = tf.Session()
    co = 0
    while (True):
        feed_dict = {x: np.expand_dims(np.arange(1000),1), 
                     col: np.random.randint( 0,900, size=( np.random.randint(500,20000),1 ) ) }
        _res = sess.run(diff, feed_dict=feed_dict)
        co = co + 1
        if (co % 1000 == 0):
            gc.collect()
            print(_res.shape)
            print(co)


def Buggy():
    with tf.device("/gpu:0"):
        x = tf.placeholder( dtype=tf.float32 , shape=(1000))
        ind = tf.placeholder( dtype=tf.int32, shape=(None) )

        p0 = tf.gather_nd(tf.expand_dims(x,1),tf.expand_dims(ind,1))
        p1 = tf.gather_nd(tf.expand_dims(x,1),tf.expand_dims(ind+1,1))

        diff = p0-p1

    sess = tf.Session()
    co = 0
    while(True):
        #Note that the length of ind is variable
        feed_dict = {x: np.arange(1000),ind: np.random.randint(0, 900, size=(np.random.randint(500, 20000)))}
        #When sized is fixed : doesn't memory leak
        #feed_dict = {x: np.arange(1000),ind: np.random.randint(0, 900, size=(20000,))}
        _res = sess.run(diff, feed_dict=feed_dict)
        co = co+1
        if( co % 1000 == 0):
            gc.collect()
            print(_res.shape)
            print(co)

#NotBuggy()
Buggy()
```

The code above leaks cpu memory quite rapidly ~10 Mb/s.
I tried some differential analysis, the bug seems weird and related to the graph construction.

Have Fun :)

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Lubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.9.0-0-g25c197e023 1.9.0
- **Python version**: Python 3.6.5 (default, May  3 2018, 10:08:28) (also bugs on Python 3.5.2 (default, Nov 17 2016, 17:05:23) )
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**:   CUDA Driver Version / Runtime Version          9.1 / 8.0 , cuDNN N/A
- **GPU model and memory**: NVidia GTx1080 ti 12Go
- **Exact command to reproduce**: Run code given above (It leaks memory quite rapidly ~10 Mb /s)

**Comment by rohan100jain** [comment]
I could reproduce the memory leak. As a guess, could this be the allocate_temp in the GatherNd op doing something unexpected?

**Comment by ebrevdo** [comment]
Just saw this. Will review tomorrow

**Comment by ebrevdo** [comment]
I don't see how `allocate_temp` on the GPU kernel would leak CPU memory, but it is possible.

**Comment by ebrevdo** [comment]
I'm able to replicate the issue.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by unrealwill** [comment]
Any update ?

**Comment by Nireyflip** [comment]
I got the same problem. I used gperf to prof the heap using status and it showed me a cpu memory increasing caused by allocate_temp, but i ran the run the funtion on a GPU kernel.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by meyerjo** [comment]
@ebrevdo An answer/update regarding this issue, but we highly appreciated. It is rather on the annoying side of bugs in my experience.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by obg2sens** [comment]
Any updates?

**Comment by ebrevdo** [comment]
I expect this bug to be gone in TF2, and TF1 is no longer officially supported by TensorFlow team.  Can you provide a TF2-based repro, one that does not use Sessions?

**Comment by unrealwill** [comment]
Yeah it has, hopefully, probably solved itself with TF2, I've moved to pytorch since.
Thanks.
